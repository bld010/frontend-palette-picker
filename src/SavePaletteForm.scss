.SavePaletteForm {
  height: auto;
  background-color: white;
  border: 1px solid lightgrey;
  border-radius: 8px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  width: 500px;
  padding: 20px;

  & .miniPalette {
    margin-left: 30px;
    width: 50%;
  }

  & h2 {
    text-align: center;
    border-bottom: 1px dotted lightgrey;
    width: 90%;
    font-family: 'Petit Formal Script', cursive;
    text-transform: lowercase;
  }



  & input:nth-of-type(1) {
    height: 40px;
    font-size: 30px;
    width: 80%;
  }
  
  & section {
    display: flex;
    flex-direction: row;
    justify-content: center;
    margin-bottom: 50px;
    
    & div {
      margin: 40px;
    }
  }
  
  & .folderList {
    box-shadow: inset 0px 0px 5px 5px rgba(0, 0, 0, 0.2);
    height: 200px;
    width: 200px;
    overflow: scroll;
    margin: 0;
    padding: 10px;

    & p {
      border-bottom: 1px dotted lightgrey;
      padding-bottom: 10px;
      cursor: pointer;
    }
  }

  & .selectFolder {
    border-right: 1px dotted lightgrey;
    padding-right: 20px;
    width: 50%;
  }

  & .createFolder, .selectFolder {
    margin: 10px;
    display: flex; 
    flex-direction: column;
    align-items: center;


    & input {
      font-size: 12px;
      width: 200px;
      height: 40px;
    }

    & button {
      margin: 10px;
      border-radius: 8px;
      font-family: 'Open Sans', sans-serif;
    }
  }

  & button {
    width: 100px;
  }

  & .active {
    background-color: lightblue;
  }

  & .bottom-buttons {
    display: flex;
    flex-direction: row;
    width: 50%;
    justify-content: center;

    & button {
      margin: 0 20px 0 20px;
      border-radius: 8px;
      font-family: 'Open Sans', sans-serif;
      font-size: 16px;

      &:nth-of-type(1) {
        background-color: lightblue;
      }
    }
  }
  .error {
    color: red
  }
}