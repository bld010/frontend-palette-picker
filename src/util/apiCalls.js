export const getFolders = async () => {
  const url = process.env.REACT_APP_BACKEND_URL + "/api/v1/folders/";
  let res = await fetch(url);
  if (!res.ok) {
    throw new Error("There was an error retrieving your folders!");
  } else {
    let data = await res.json();
    return data;
  }
};

export const getPalettes = async () => {
  const url = process.env.REACT_APP_BACKEND_URL + "/api/v1/palettes/";
  let res = await fetch(url);
  if (!res.ok) {
    throw new Error("There was an error retrieving your palettes!");
  } else {
    let data = await res.json();
    return data;
  }
};

export const deleteFolder = async id => {
  const url = process.env.REACT_APP_BACKEND_URL + `/api/v1/folders/${id}`;
  const options = {
    method: "DELETE",
    headers: { "Content-Type": "application/json" }
  };
  try {
    const res = await fetch(url, options);
    if (!res.ok) {
      throw new Error("Cannot delete folder!");
    }
    const deletedFolder = await res.json();
    return deletedFolder;
  } catch (error) {
    throw new Error(error);
  }
};

export const deletePalette = async id => {
  const url = process.env.REACT_APP_BACKEND_URL + `/api/v1/palettes/${id}`;
  const options = {
    method: "DELETE",
    headers: { "Content-Type": "application/json" }
  };
  try {
    const res = await fetch(url, options);
    if (!res.ok) {
      throw new Error("Cannot delete palette!");
    }
    const deletedPalette = await res.json();
    return deletedPalette;
  } catch (error) {
    throw new Error(error);
  }
};

export const postFolder = async folderName => {
  const url = process.env.REACT_APP_BACKEND_URL + `/api/v1/folders/`;
  const body = {
    name: folderName
  };
  const options = {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body)
  };

  try {
    const res = await fetch(url, options);
    if (!res.ok) {
      throw new Error("There was an error posting this folder!");
    }
    const newFolder = await res.json();
    return newFolder;
  } catch (error) {
    throw new Error(error);
  }
};

export const postPalette = async (
  colorOne,
  colorTwo,
  colorThree,
  colorFour,
  colorFive,
  folderId,
  paletteName
) => {
  const url = process.env.REACT_APP_BACKEND_URL + `/api/v1/palettes/`;
  const body = {
    color1: colorOne,
    color2: colorTwo,
    color3: colorThree,
    color4: colorFour,
    color5: colorFive,
    folder_id: folderId,
    name: paletteName
  };
  const options = {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(body)
  };

  try {
    const res = await fetch(url, options);
    if (!res.ok) {
      throw new Error("There was an error posting this palette!");
    }
    const newPalette = await res.json();
    return newPalette;
  } catch (error) {
    throw new Error(error);
  }
};

// export const patchFolder = async (folderName, id) => {
//   const url = process.env.REACT_APP_BACKEND_URL + `/api/v1/folders/${id}`;
//   const body = {
//     name: folderName
//   };
//   const options = {
//     method: "PATCH",
//     headers: { "Content-Type": "application/json" },
//     body: JSON.stringify(body)
//   };

//   try {
//     const res = await fetch(url, options);
//     if (!res.ok) {
//       const error = await res.json();
//       throw new Error(error);
//     }
//     const patchedFolder = await res.json();
//     return patchedFolder;
//   } catch (error) {
//     throw new Error(error);
//   }
// };

// export const patchPalette = async (
//   colorOne,
//   colorTwo,
//   colorThree,
//   colorFour,
//   colorFive,
//   folderId,
//   paletteName,
//   id
// ) => {
//   const url = process.env.REACT_APP_BACKEND_URL + `/api/v1/palettes/${id}`;
//   const body = {
//     color1: colorOne,
//     color2: colorTwo,
//     color3: colorThree,
//     color4: colorFour,
//     color5: colorFive,
//     folder_id: folderId,
//     name: paletteName
//   };
//   const options = {
//     method: "PATCH",
//     headers: { "Content-Type": "application/json" },
//     body: JSON.stringify(body)
//   };

//   try {
//     const res = await fetch(url, options);
//     if (!res.ok) {
//       const error = await res.json();
//       throw new Error(error);
//     }
//     const patchedPalette = await res.json();
//     return patchedPalette;
//   } catch (error) {
//     throw new Error(error);
//   }
// };
