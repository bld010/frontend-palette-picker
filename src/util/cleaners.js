export const cleanPalettes = rawPalettes => {
  return rawPalettes.map(palette => {
    let {
      name,
      colors,
      id,
      folder_id,
      color1,
      color2,
      color3,
      color4,
      color5
    } = palette;

    let cleanedPalette = {
      name,
      id,
      folder_id,
      colors: [color1, color2, color3, color4, color5]
    };

    return cleanedPalette;
  });
};

export const cleanFolders = rawFolders => {
  return rawFolders.map(folder => {
    let { id, name } = folder;
    let cleanedFolder = { id, name, palettes: [] };
    return cleanedFolder;
  });
};

export const cleanData = (folders, palettes) => {
  return folders.map(folder => {
    palettes.forEach(palette => {
      if (palette.folder_id === folder.id) {
        folder.palettes.push(palette);
      }
    });
    return folder;
  });
};
