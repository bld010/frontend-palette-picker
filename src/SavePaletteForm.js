import React, {Component} from 'react';
import { postFolder, getFolders } from './util/apiCalls';
import './SavePaletteForm.scss';
import MiniPalette from './MiniPalette';

export class SavePaletteForm extends Component {
    constructor(props) {
      super(props);
      this.state = {
          currentFolder: null,
          paletteName: '',
          folderName: '',
          reload: false,
          folders: this.props.folders || null
      };
    }

    handleChangeOfInput = e => {
        e.preventDefault()
        this.setState({[e.target.name]: e.target.value });
      };

    createNewFolder = async e => {
        e.preventDefault()
        await this.setState({currentFolder: this.state.folderName})
        await postFolder(this.state.folderName)
        let allFolders = await getFolders()
        await this.setState({folders: allFolders})
    }

    setCurrentFolderByClick = (e, folder) => {
        e.preventDefault()
        this.setState({currentFolder: folder})
    }

    displayFolders = () => {
        return this.state.folders.map(folder => {
            return (
                <p className={this.state.currentFolder === folder ? 'active' : ' '}
                    onClick={(e) => this.setState({currentFolder: folder})}>{folder.name}</p> 
            )
        })
    }


    componentDidUpdate = (prevProps) => {
        if (this.props.folders !== prevProps.folders) {
          this.setState( { reload: true })
        }
      }

    

    render() {
        return (
        <form className="SavePaletteForm">
            <h2>Save Your Palette</h2>


            <input type='text' placeholder='New palette name' name='paletteName' value={this.state.paletteName} onChange={e => this.handleChangeOfInput(e)}/>
            <div className="miniPalette"> 
                <MiniPalette palette={this.props.palette} />
            </div>
            <section>
                <div className="selectFolder">
                    <h3>Select A Folder</h3>
                        <div className="folderList">
                            {this.displayFolders()} 
                        </div>
                </div>
                <div className="createFolder">
                    <h3>Create a New Folder</h3>
                    <input type='text' placeholder='New folder name' name='folderName' value={this.state.folderName} onChange={e => this.handleChangeOfInput(e)}/>
                    <button onClick={this.createNewFolder}>Create Folder</button>
                   
                </div>
            </section>
            <div className="bottom-buttons">
                <button onClick={(e) => this.props.savePalette(e,this.state.currentFolder, this.state.paletteName)}>Submit!</button>
                <button onClick={this.props.hideModal}>Cancel</button>
            </div>
        </form>
        )
    }
}

export default SavePaletteForm;