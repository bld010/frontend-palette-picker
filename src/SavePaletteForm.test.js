import React from "react";
import { shallow } from "enzyme";
import SavePaletteForm from "./SavePaletteForm";
import { configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { postFolder, getFolders } from './util/apiCalls';

jest.mock('./util/apiCalls.js')

configure({ adapter: new Adapter() });

describe("SavePaletteForm", () => {
  let mockHideModal = jest.fn();
  let mockSavePalette = jest.fn();
  let mockFolders = [
    {
      id: 84,
      name: "Almost Blues",
      created_at: "2019-10-15T01:01:13.675Z",
      updated_at: "2019-10-15T01:01:13.675Z"
    },
    {
      id: 85,
      name: "Fuzzy Purple",
      created_at: "2019-10-15T01:01:50.370Z",
      updated_at: "2019-10-15T01:01:50.370Z"
    },
    {
      id: 86,
      name: "Rubber Ducky Takes a Bath",
      created_at: "2019-10-15T01:01:13.675Z",
      updated_at: "2019-10-15T01:01:13.675Z"
    },
    {
      id: 87,
      name: "Money Greens",
      created_at: "2019-10-15T01:01:50.370Z",
      updated_at: "2019-10-15T01:01:50.370Z"
    }
  ];

  let mockCurrentPalette = {
    id: 125,
    color1: "#00ed3f",
    color2: "#e09e72",
    color3: "#166c8e",
    color4: "#d1ce25",
    color5: "#f9dd89",
    name: "1",
    folder_id: 84,
    created_at: "2019-10-15T01:01:16.495Z",
    updated_at: "2019-10-15T01:01:16.495Z"
  };

  let wrapper = shallow(
    <SavePaletteForm
      hideModal={mockHideModal}
      savePalette={mockSavePalette}
      folders={mockFolders}
      palette={mockCurrentPalette}
    />
  );

  it("should match the snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should update state when handleChangeOfInput is called", () => {
    const mockEvent = {
      target: { name: "folderName", value: "Sweet Potatoes" },
      preventDefault: jest.fn()
    };
    const expected = "Sweet Potatoes";

    wrapper.instance().handleChangeOfInput(mockEvent);

    expect(wrapper.state("folderName")).toEqual(expected);
  });

  it("should update state with new folder when createNewFolders is called", async () => {
    const mockEvent = { preventDefault: jest.fn() };

    let mockNewFolder = {id: 98, name: 'New Folder'}

    wrapper.instance().setState({ folderName: 'New Folder'});

    postFolder.mockImplementation(() => {
      return Promise.resolve()
    })

    let mockNewFolders = [
      {id: 84, name: 'Almost Blues'},
      mockNewFolder
    ]

    getFolders.mockImplementation(() => {
      return Promise.resolve(mockNewFolders)
    })

    await wrapper.instance().createNewFolder(mockEvent);

    expect(wrapper.state().currentFolder).toEqual(mockNewFolder);
    expect(wrapper.state().folders).toEqual(mockNewFolders)
  });

  it('should update state with an error if name is already taken', async () => {
    const mockEvent = { preventDefault: jest.fn() };


    wrapper.instance().setState({ folderName: 'Almost Blues'});

    await wrapper.instance().createNewFolder(mockEvent);

    expect(wrapper.state().error).toEqual("This folder already exists! Please choose a new name!")
  })

  it("should fire savePallete when the currentFolder and paletteName are true in state", () => {
    let mockEvent = { preventDefault: jest.fn() };
    let defaultState = { currentFolder: true, paletteName: true };

    wrapper.instance().setState(defaultState);
    wrapper.instance().handleSave(mockEvent);

    expect(mockSavePalette).toHaveBeenCalled();
  });

  it("should update state with an error when a paletteName field is missing", () => {
    let mockEvent = { preventDefault: jest.fn() };
    let defaultState = { currentFolder: true, paletteName: false };

    wrapper.instance().setState(defaultState);
    wrapper.instance().handleSave(mockEvent);
    expect(wrapper.state().error).toEqual(
      "Please enter a name for a palette and select a folder!"
    );
  });

  it("should update state with an error when a currentFolder field is missing", () => {
    let mockEvent = { preventDefault: jest.fn() };
    let defaultState = { currentFolder: false, paletteName: true };

    wrapper.instance().setState(defaultState);
    wrapper.instance().handleSave(mockEvent);
    expect(wrapper.state().error).toEqual(
      "Please enter a name for a palette and select a folder!"
    );
  });

  it("should set currentFolder on a click", () => {
    wrapper
      .find(".folderList")
      .childAt(0)
      .simulate("click");
    expect(wrapper.state().currentFolder.name).toEqual("Almost Blues");
  });
});
