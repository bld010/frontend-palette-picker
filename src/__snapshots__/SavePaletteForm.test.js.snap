// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SavePaletteForm should match the snapshot 1`] = `
<form
  className="SavePaletteForm"
>
  <h2>
    Save Your Palette
  </h2>
  <input
    name="paletteName"
    onChange={[Function]}
    placeholder="New palette name"
    type="text"
    value=""
  />
  <div
    className="miniPalette"
  >
    <MiniPalette
      palette={
        Object {
          "color1": "#00ed3f",
          "color2": "#e09e72",
          "color3": "#166c8e",
          "color4": "#d1ce25",
          "color5": "#f9dd89",
          "created_at": "2019-10-15T01:01:16.495Z",
          "folder_id": 84,
          "id": 125,
          "name": "1",
          "updated_at": "2019-10-15T01:01:16.495Z",
        }
      }
    />
  </div>
  <section>
    <div
      className="selectFolder"
    >
      <h3>
        Select A Folder
      </h3>
      <div
        className="folderList"
      >
        <p
          className=" "
          key="84"
          onClick={[Function]}
        >
          Almost Blues
        </p>
        <p
          className=" "
          key="85"
          onClick={[Function]}
        >
          Fuzzy Purple
        </p>
        <p
          className=" "
          key="86"
          onClick={[Function]}
        >
          Rubber Ducky Takes a Bath
        </p>
        <p
          className=" "
          key="87"
          onClick={[Function]}
        >
          Money Greens
        </p>
      </div>
    </div>
    <div
      className="createFolder"
    >
      <h3>
        Create a New Folder
      </h3>
      <input
        name="folderName"
        onChange={[Function]}
        placeholder="New folder name"
        type="text"
        value=""
      />
      <button
        onClick={[Function]}
      >
        Create Folder
      </button>
    </div>
  </section>
  <div
    className="bottom-buttons"
  >
    <button
      onClick={[Function]}
    >
      Submit!
    </button>
    <button
      onClick={[MockFunction]}
    >
      Cancel
    </button>
  </div>
</form>
`;
